services/api/lib/postgre.js

```diff
@@ -295,6 +295,334 @@ module.exports = function (pg) {
+            if (
+              result.method === 'create' &&
+              result.type === lookupType &&
+              result.id === lookupId )
+            {
+              txResult = result;
+              break;
+            }
+          }
+
+          // this action refers to an object created earlier in the transaction
+          if ( txResult ) {
+            return resolve(txResult.user_id === userId);
+          }
+
+          // look up the parent type for this action, to verify the user has ownership
+          executeTransaction(transaction, queries[lookupType].findOneById, [lookupId])
+          .then(function(results) {
+            if ( !results.rows.length || results.rows[0].user_id !== userId ) {
+              return resolve(false);
+            }
+            resolve(true);
+          })
+          .catch(function(err) {
+            reject(err);
+          });
+        });
+      }
+
+      server.method('projects.bulk', function(actions, userId, done) {
+        var transaction;
+        var transactionResults;
+        var failureData;
+        var errorReason;
+
+        // reduce the Array of actions, building a new array of transaction results
+        function reduceActions(results, action, actionIndex) {
+
+          // check if a key on the action object should be resolved to a value
+          // returned by a previous action in the transaction
+          function everyActionKey(key) {
+            // if there's a problem resolving the key to a value, this is set to false
+            var valid = true;
+
+            // grab the index ($2) of the action results to reach into, and grab the value described by reachString ($3) using Hoek.reach()
+            function replace(match, $2, $3) {
+              var reachString = $3;
+              var reachIdx = +$2;
+              var value;
+
+              if ( reachIdx <= results.length ) {
+                value = Hoek.reach(results[reachIdx], reachString);
+              } else {
+                errorReason = 'Array reference out of bounds for ' + key + ' in action at index ' + actionIndex;
+                failureData = {
+                  key: key,
+                  reachIdx: reachIdx,
+                  actionIndex: actionIndex
+                };
+                valid = false;
+                return;
+              }
+
+              if ( !value ) {
+                errorReason = 'Invalid reference to value using key \'' +
+                  key +
+                  '\' in action at index ' +
+                  actionIndex;
+                failureData = {
+                  key: key,
+                  reachIdx: reachIdx,
+                  actionIndex: actionIndex
+                };
+                valid = false;
+                return;
+              }
+
+              return value;
+            }
+
+            // if the key's value isn't a reach string, return valid
+            if ( !reachRegex.test(action.data[key]) ) {
```

Comment:
Since we're returning `valid` in both cases, we can compact this to:
```
if ( reachRegex.test(action.data[key]) ) {
  action.data[key] = action.data[key].replace(reachRegex, replace);
}
return valid;
```
unless that replace() call modified `valid`, which is probably not a good idea.